src/lib/features/client-feature-toggles/delta/client-feature-toggle-delta.ts

```diff
@@ -18,92 +18,75 @@ import { CLIENT_DELTA_MEMORY } from '../../../metric-events';
-};
-
-export type RevisionDeltaEntry = {
-    updated: FeatureConfigurationDeltaClient[];
-    revisionId: number;
-    removed: DeletedFeature[];
-    segments: IClientSegment[];
-};
-
-export type Revision = {
-    revisionId: number;
-    updated: any[];
-    removed: DeletedFeature[];
-};
-
-type Revisions = Record<string, RevisionDelta>;
-
-const applyRevision = (first: Revision, last: Revision): Revision => {
-    const updatedMap = new Map(
-        [...first.updated, ...last.updated].map((feature) => [
-            feature.name,
-            feature,
-        ]),
-    );
-    const removedMap = new Map(
-        [...first.removed, ...last.removed].map((feature) => [
-            feature.name,
-            feature,
-        ]),
+import {
+    DELTA_EVENT_TYPES,
+    type DeltaEvent,
+    type DeltaHydrationEvent,
+    isDeltaFeatureRemovedEvent,
+    isDeltaFeatureUpdatedEvent,
+} from './client-feature-toggle-delta-types';
+
+type EnvironmentRevisions = Record<string, DeltaCache>;
+
+export const filterEventsByQuery = (
+    events: DeltaEvent[],
+    requiredRevisionId: number,
+    projects: string[],
+    namePrefix: string,
+) => {
+    const targetedEvents = events.filter(
+        (revision) => revision.eventId > requiredRevisionId,
     );
-
-    for (const feature of last.updated) {
-        removedMap.delete(feature.name);
-    }
-
-    for (const feature of last.removed) {
-        updatedMap.delete(feature.name);
-        removedMap.set(feature.name, feature);
-    }
-
-    return {
-        revisionId: last.revisionId,
-        updated: Array.from(updatedMap.values()),
-        removed: Array.from(removedMap.values()),
+    const allProjects = projects.includes('*');
+    const startsWithPrefix = (revision: DeltaEvent) => {
+        return (
+            (isDeltaFeatureUpdatedEvent(revision) &&
+                revision.feature.name.startsWith(namePrefix)) ||
+            (isDeltaFeatureRemovedEvent(revision) &&
+                revision.featureName.startsWith(namePrefix))
+        );
     };
-};
 
-const filterRevisionByProject = (
-    revision: Revision,
-    projects: string[],
-): Revision => {
-    const updated = revision.updated.filter(
-        (feature) =>
-            projects.includes('*') || projects.includes(feature.project),
-    );
-    const removed = revision.removed.filter(
-        (feature) =>
-            projects.includes('*') || projects.includes(feature.project),
-    );
+    const isInProject = (revision: DeltaEvent) => {
+        return (
+            (isDeltaFeatureUpdatedEvent(revision) &&
+                projects.includes(revision.feature.project!)) ||
+            (isDeltaFeatureRemovedEvent(revision) &&
+                projects.includes(revision.project))
+        );
+    };
 
-    return { ...revision, updated, removed };
+    return targetedEvents.filter((revision) => {
+        return (
+            startsWithPrefix(revision) && (allProjects || isInProject(revision))
+        );
```

Comment:
We always filter with namePrefix? What if nameprefix is empty?

Reply:
https://github.com/Unleash/unleash/pull/9133/files#diff-117921481602afc7d7b53fef3f43e7c22c8e08d938f28512444b1021bdcb86c5R145

Normalizes it.